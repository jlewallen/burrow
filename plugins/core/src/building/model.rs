use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
