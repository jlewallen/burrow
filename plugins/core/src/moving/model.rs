use crate::library::model::*;

pub trait BeforeMovingHook {
    fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove>;
}

impl BeforeMovingHook for MovingHooks {
    fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove> {
        Ok(self
            .before_moving
            .instances
            .borrow()
            .iter()
            .map(|h| h.before_moving(surroundings, to_area))
            .collect::<Result<Vec<CanMove>>>()?
            .iter()
            .fold(CanMove::default(), |c, h| c.fold(h)))
    }
}

pub trait AfterMoveHook {
    fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()>;
}

impl AfterMoveHook for MovingHooks {
    fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()> {
        self.after_move
            .instances
            .borrow()
            .iter()
            .map(|h| h.after_move(surroundings, from_area))
            .collect::<Result<Vec<()>>>()?;

        Ok(())
    }
}

#[derive(Default)]
pub struct MovingHooks {
    pub before_moving: Hooks<Box<dyn BeforeMovingHook>>,
    pub after_move: Hooks<Box<dyn AfterMoveHook>>,
}

impl HooksSet for MovingHooks {
    fn hooks_key() -> &'static str
    where
        Self: Sized,
    {
        "moving"
    }
}

#[derive(Clone, Default)]
pub enum CanMove {
    #[default]
    Allow,
    Prevent,
}

impl HookOutcome for CanMove {
    fn fold(&self, other: &Self) -> Self {
        match (self, other) {
            (_, CanMove::Prevent) => CanMove::Prevent,
            (CanMove::Prevent, _) => CanMove::Prevent,
            (_, _) => CanMove::Allow,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Occupying {
    pub area: EntityRef,
}

impl Scope for Occupying {
    fn scope_key() -> &'static str {
        "occupying"
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SimpleRoute {
    name: String,
    to: EntityRef,
}

impl SimpleRoute {
    pub fn new(name: &str, to: EntityRef) -> Self {
        Self {
            name: name.to_owned(),
            to,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
#[serde(rename_all = "camelCase")]
pub enum Route {
    Simple(SimpleRoute),
}

impl Route {
    fn conflicts_with(&self, other: &Route) -> bool {
        match (self, other) {
            (Route::Simple(a), Route::Simple(b)) => a.name == b.name,
        }
    }

    fn matching_name(&self, name: &str) -> bool {
        match self {
            Route::Simple(simple) => simple.name.to_lowercase().contains(&name.to_lowercase()),
        }
    }

    pub fn destination(&self) -> &EntityRef {
        match self {
            Route::Simple(simple) => &simple.to,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Occupyable {
    pub acls: Acls,
    pub occupied: Vec<EntityRef>,
    pub occupancy: u32,
    pub routes: Option<Vec<Route>>,
}

impl Occupyable {
    pub fn stop_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
        let before = self.occupied.len();
        self.occupied.retain(|i| *i.key() != item.key());
        let after = self.occupied.len();
        if before == after {
            return Ok(DomainOutcome::Nope);
        }

        Ok(DomainOutcome::Ok)
    }

    pub fn start_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
        self.occupied.push(item.entity_ref());

        Ok(DomainOutcome::Ok)
    }

    pub fn remove_route(&mut self, name: &str) -> Result<bool> {
        if let Some(routes) = &mut self.routes {
            if let Some(found) = routes.iter().position(|r| r.matching_name(name)) {
                routes.remove(found);
                return Ok(true);
            }
        }

        Ok(false)
    }

    pub fn add_route(&mut self, route: Route) -> Result<(), DomainError> {
        let routes = self.routes.get_or_insert_with(|| Vec::new());

        if let Some(conflict) = routes.iter().position(|r| r.conflicts_with(&route)) {
            routes.remove(conflict);
        }

        routes.push(route);

        Ok(())
    }

    pub fn find_route(&self, item: &Item) -> Result<Option<EntityPtr>, DomainError> {
        let Some(routes) = &self.routes else {
            return Ok(None);
        };

        match item {
            Item::Route(name) => {
                for route in routes {
                    if route.matching_name(&name) {
                        return Ok(Some(route.destination().to_entity()?));
                    }
                }
                Ok(None)
            }
            _ => panic!("Occupyable::find_route expecting Item::Route"),
        }
    }
}

impl Scope for Occupyable {
    fn scope_key() -> &'static str {
        "occupyable"
    }
}
