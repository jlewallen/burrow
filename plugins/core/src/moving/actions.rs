use std::rc::Rc;

use crate::library::actions::*;
use crate::looking::actions::*;
use crate::looking::model::Observe;
use crate::moving::model::Route;

use super::model::Occupyable;

#[action]
pub struct GoAction {
    pub item: Item,
}

impl GoAction {
    fn navigate(
        &self,
        session: SessionRef,
        actor: EntityPtr,
        area: EntityPtr,
        to_area: EntityPtr,
    ) -> ReplyResult {
        match tools::navigate_between(&area, &to_area, &actor)? {
            true => {
                let excluding = actor.key();
                let hearing_arrive: Vec<_> = tools::get_occupant_keys(&to_area)?
                    .into_iter()
                    .filter(|v| *v != excluding)
                    .collect();

                session.raise(
                    Some(actor.clone()),
                    Audience::Area(area.key().clone()),
                    Raising::TaggedJson(
                        Moving::Left {
                            actor: (&actor).observe(&actor)?.expect("No observed entity"),
                            area: (&area).observe(&actor)?.expect("No observed entity"),
                        }
                        .to_tagged_json()?,
                    ),
                )?;
                session.raise(
                    Some(actor.clone()),
                    Audience::Individuals(hearing_arrive),
                    Raising::TaggedJson(
                        Moving::Arrived {
                            actor: (&actor).observe(&actor)?.expect("No observed entity"),
                            area: (&to_area).observe(&actor)?.expect("No observed entity"),
                        }
                        .to_tagged_json()?,
                    ),
                )?;

                Ok(session.perform(Perform::Actor {
                    actor,
                    action: PerformAction::Instance(Rc::new(LookAction {})),
                })?)
            }
            false => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

impl Action for GoAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("go {:?}!", self.item);

        let (_, actor, area) = surroundings.unpack();

        if let Some(occupyable) = area.scope::<Occupyable>()? {
            match &self.item {
                Item::Route(route) => match occupyable.find_route(&route) {
                    Some(route) => match route {
                        Route::Simple(to_area) => {
                            let to_area = to_area.destination().to_entity()?;
                            self.navigate(session, actor, area, to_area)
                        }
                        Route::Deactivated(reason, _) => {
                            Ok(SimpleReply::Prevented(Some(reason.clone())).try_into()?)
                        }
                    },
                    None => {
                        match session.find_item(surroundings, &Item::Named(route.to_owned()))? {
                            Some(maybe) => {
                                if maybe.scope::<Occupyable>()?.is_some() {
                                    self.navigate(session, actor, area, maybe)
                                } else {
                                    Ok(SimpleReply::NotFound.try_into()?)
                                }
                            }
                            None => Ok(SimpleReply::NotFound.try_into()?),
                        }
                    }
                },
                Item::Gid(_) => match session.find_item(surroundings, &self.item)? {
                    Some(to_area) => self.navigate(session, actor, area, to_area),
                    None => Ok(SimpleReply::NotFound.try_into()?),
                },
                _ => panic!("Occupyable::find_route expecting Item::Route or Item::Gid"),
            }
        } else {
            Ok(SimpleReply::NotFound.try_into()?)
        }
    }
}

#[action]
pub struct ShowRoutesAction {}

impl Action for ShowRoutesAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        let (_, _, area) = surroundings.unpack();
        let Some(occupyable) = area.scope::<Occupyable>()? else {
            return Ok(SimpleReply::NotFound.try_into()?);
        };
        let Some(routes) = &occupyable.routes else {
            return Ok(SimpleReply::NotFound.try_into()?);
        };

        let reply = TaggedJson::new("routes".to_owned(), serde_json::to_value(routes)?.into());

        Ok(Effect::Reply(EffectReply::TaggedJson(reply)))
    }
}

#[action]
pub struct AddRouteAction {
    pub area: Item,
    pub name: String,
    pub destination: Item,
}

impl Action for AddRouteAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        match session.find_item(&surroundings, &self.area)? {
            Some(area) => {
                let Some(destination) = session.find_item(surroundings, &self.destination)? else {
                    return Ok(SimpleReply::NotFound.try_into()?);
                };

                tools::add_route(&area, &self.name, &destination)?;

                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::Done.try_into()?),
        }
    }
}

#[action]
pub struct RemoveRouteAction {
    pub area: Item,
    pub name: String,
}

impl Action for RemoveRouteAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        match session.find_item(&surroundings, &self.area)? {
            Some(area) => {
                let mut occupyable = area.scope_mut::<Occupyable>()?;
                if !occupyable.remove_route(&self.name) {
                    return Ok(SimpleReply::NotFound.try_into()?);
                }

                occupyable.save()?;

                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct ActivateRouteAction {
    pub name: String,
}

impl Action for ActivateRouteAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        let (_, _, area) = surroundings.unpack();

        let mut occupyable = area.scope_mut::<Occupyable>()?;
        occupyable.activate(&self.name);
        occupyable.save()?;

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct DeactivateRouteAction {
    pub name: String,
    pub reason: String,
}

impl Action for DeactivateRouteAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        let (_, _, area) = surroundings.unpack();

        let mut occupyable = area.scope_mut::<Occupyable>()?;
        occupyable.deactivate(&self.name, &self.reason);
        occupyable.save()?;

        Ok(SimpleReply::Done.try_into()?)
    }
}
