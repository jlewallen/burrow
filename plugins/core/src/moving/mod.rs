use crate::library::plugin::*;

#[cfg(test)]
mod tests;

#[derive(Default)]
pub struct MovingPluginFactory {}

impl PluginFactory for MovingPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(MovingPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct MovingPlugin {}

impl Plugin for MovingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "moving"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn initialize(&mut self) -> Result<()> {
        Ok(())
    }

    fn middleware(&mut self) -> Result<Vec<Rc<dyn Middleware>>> {
        Ok(Vec::default())
    }

    fn deliver(&self, _incoming: &Incoming) -> Result<()> {
        Ok(())
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

impl ParsesActions for MovingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::GoActionParser {}, i)
    }
}

pub mod model {
    use crate::library::model::*;

    pub trait BeforeMovingHook {
        fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove>;
    }

    impl BeforeMovingHook for MovingHooks {
        fn before_moving(&self, surroundings: &Surroundings, to_area: &EntityPtr) -> Result<CanMove> {
            Ok(self
                .before_moving
                .instances
                .borrow()
                .iter()
                .map(|h| h.before_moving(surroundings, to_area))
                .collect::<Result<Vec<CanMove>>>()?
                .iter()
                .fold(CanMove::default(), |c, h| c.fold(h)))
        }
    }

    pub trait AfterMoveHook {
        fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()>;
    }

    impl AfterMoveHook for MovingHooks {
        fn after_move(&self, surroundings: &Surroundings, from_area: &EntityPtr) -> Result<()> {
            self.after_move
                .instances
                .borrow()
                .iter()
                .map(|h| h.after_move(surroundings, from_area))
                .collect::<Result<Vec<()>>>()?;

            Ok(())
        }
    }

    #[derive(Default)]
    pub struct MovingHooks {
        pub before_moving: Hooks<Box<dyn BeforeMovingHook>>,
        pub after_move: Hooks<Box<dyn AfterMoveHook>>,
    }

    impl HooksSet for MovingHooks {
        fn hooks_key() -> &'static str
        where
            Self: Sized,
        {
            "moving"
        }
    }

    #[derive(Clone, Default)]
    pub enum CanMove {
        #[default]
        Allow,
        Prevent,
    }

    impl HookOutcome for CanMove {
        fn fold(&self, other: &Self) -> Self {
            match (self, other) {
                (_, CanMove::Prevent) => CanMove::Prevent,
                (CanMove::Prevent, _) => CanMove::Prevent,
                (_, _) => CanMove::Allow,
            }
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Occupying {
        pub area: EntityRef,
    }

    impl Scope for Occupying {
        fn scope_key() -> &'static str {
            "occupying"
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Occupyable {
        pub acls: Acls,
        pub occupied: Vec<EntityRef>,
        pub occupancy: u32,
    }

    impl Occupyable {
        pub fn stop_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
            let before = self.occupied.len();
            self.occupied.retain(|i| *i.key() != item.key());
            let after = self.occupied.len();
            if before == after {
                return Ok(DomainOutcome::Nope);
            }

            Ok(DomainOutcome::Ok)
        }

        pub fn start_occupying(&mut self, item: &EntityPtr) -> Result<DomainOutcome> {
            self.occupied.push(item.entity_ref());

            Ok(DomainOutcome::Ok)
        }
    }

    impl Scope for Occupyable {
        fn scope_key() -> &'static str {
            "occupyable"
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Exit {
        pub area: EntityRef,
    }

    impl Scope for Exit {
        fn scope_key() -> &'static str {
            "exit"
        }
    }

    #[derive(Debug, Serialize, Deserialize)]
    pub struct AreaRoute {
        pub area: EntityRef,
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Movement {
        pub routes: Vec<AreaRoute>,
    }

    impl Scope for Movement {
        fn scope_key() -> &'static str {
            "movement"
        }
    }
}

pub mod actions {
    use crate::library::actions::*;
    use crate::looking::actions::*;
    use crate::looking::model::Observe;
    use crate::moving::model::{AfterMoveHook, BeforeMovingHook, CanMove, MovingHooks};

    #[action]
    pub struct GoAction {
        pub item: Item,
    }

    impl Action for GoAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("go {:?}!", self.item);

            let (_, living, area) = surroundings.unpack();

            match session.find_item(surroundings, &self.item)? {
                Some(to_area) => {
                    let can = session.hooks().invoke::<MovingHooks, CanMove, _>(|h| {
                        h.before_moving(surroundings, &to_area)
                    })?;

                    match can {
                        CanMove::Allow => {
                            match tools::navigate_between(&area, &to_area, &living)? {
                                DomainOutcome::Ok => {
                                    session.hooks().invoke::<MovingHooks, (), _>(|h| {
                                        h.after_move(surroundings, &area)
                                    })?;

                                    session.raise(
                                        Audience::Area(area.key().clone()),
                                        Raising::TaggedJson(
                                            MovingEvent::Left {
                                                living: (&living)
                                                    .observe(&living)?
                                                    .expect("No observed entity"),
                                                area: (&area)
                                                    .observe(&living)?
                                                    .expect("No observed entity"),
                                            }
                                            .to_tagged_json()?,
                                        ),
                                    )?;
                                    session.raise(
                                        Audience::Area(to_area.key().clone()),
                                        Raising::TaggedJson(
                                            MovingEvent::Arrived {
                                                living: (&living)
                                                    .observe(&living)?
                                                    .expect("No observed entity"),
                                                area: (&to_area)
                                                    .observe(&living)?
                                                    .expect("No observed entity"),
                                            }
                                            .to_tagged_json()?,
                                        ),
                                    )?;

                                    session.perform(Perform::Living {
                                        living,
                                        action: PerformAction::Instance(Rc::new(LookAction {})),
                                    })
                                }
                                DomainOutcome::Nope => Ok(SimpleReply::NotFound.try_into()?),
                            }
                        }
                        CanMove::Prevent => Ok(SimpleReply::Prevented.try_into()?),
                    }
                }
                None => Ok(SimpleReply::NotFound.try_into()?),
            }
        }
    }
}

mod parser {
    use crate::library::parser::*;

    use super::actions::GoAction;

    pub struct GoActionParser {}

    impl ParsesActions for GoActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let (_, action) = map(
                separated_pair(tag("go"), spaces, named_place),
                |(_, target)| GoAction { item: target },
            )(i)?;

            Ok(Some(Box::new(action)))
        }
    }
}
