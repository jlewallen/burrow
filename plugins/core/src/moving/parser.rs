use crate::library::parser::*;

use super::actions::GoAction;

pub struct GoActionParser {}

impl ParsesActions for GoActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            separated_pair(tag("go"), spaces, named_place),
            |(_, target)| GoAction { item: target },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
