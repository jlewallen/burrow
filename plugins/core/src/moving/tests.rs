use super::parser::*;
use super::*;
use crate::library::tests::*;
use crate::looking::model::new_area_observation;
use crate::moving::actions::{
    AddRouteAction, DeactivateRouteAction, RemoveRouteAction, ShowRoutesAction,
};
use crate::moving::model::{Occupyable, Route, SimpleRoute};

#[test]
fn it_goes_ignores_bad_matches() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let east = build.make(QuickThing::Place("East Place"))?;
    let west = build.make(QuickThing::Place("West Place"))?;
    let (session, surroundings) = build
        .route("East", QuickThing::Actual(east))
        .route("Wast", QuickThing::Actual(west))
        .build()?;

    let action = try_parsing(GoActionParser {}, "go north")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    build.close()?;

    Ok(())
}

#[test]
fn it_goes_through_correct_route_when_two_nearby() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let east = build.make(QuickThing::Place("East Place"))?;
    let west = build.make(QuickThing::Place("West Place"))?;
    let (session, surroundings) = build
        .route("East", QuickThing::Actual(east.clone()))
        .route("Wast", QuickThing::Actual(west))
        .build()?;

    let action = try_parsing(GoActionParser {}, "go east")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, area) = surroundings.unpack();

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &east)?);

    assert_ne!(tools::area_of(&living)?.key(), area.key());
    assert_eq!(tools::area_of(&living)?.key(), east.key());

    build.close()?;

    Ok(())
}

#[test]
fn it_goes_through_routes_when_one_nearby() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let (session, surroundings) = build
        .route("East", QuickThing::Actual(destination.clone()))
        .build()?;

    let action = try_parsing(GoActionParser {}, "go east")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, area) = surroundings.unpack();

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &destination)?);

    assert_ne!(tools::area_of(&living)?.key(), area.key());
    assert_eq!(tools::area_of(&living)?.key(), destination.key());

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_go_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(GoActionParser {}, "go rake")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_go_non_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Rake")])
        .build()?;

    let action = try_parsing(GoActionParser {}, "go rake")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    build.close()?;

    Ok(())
}

#[test]
fn it_parses_show_routes() -> Result<()> {
    let action = try_parsing(RouteActionParser {}, "@route")?;
    assert_eq!(
        action.unwrap().to_tagged_json()?,
        ShowRoutesAction {}.to_tagged_json()?
    );

    Ok(())
}

#[test]
fn it_parses_add_or_set_route() -> Result<()> {
    let action = try_parsing(RouteActionParser {}, "@route #5 north")?;
    assert_eq!(
        action.unwrap().to_tagged_json()?,
        AddRouteAction {
            area: Item::Area,
            name: "north".to_owned(),
            destination: Item::Gid(EntityGid::new(5)),
        }
        .to_tagged_json()?
    );

    Ok(())
}

#[test]
fn it_parses_remove_route() -> Result<()> {
    let action = try_parsing(RouteActionParser {}, "@route rm north")?;
    assert_eq!(
        action.unwrap().to_tagged_json()?,
        RemoveRouteAction {
            area: Item::Area,
            name: "north".to_owned()
        }
        .to_tagged_json()?
    );

    Ok(())
}

#[test]
fn it_adds_simple_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", gid))?.unwrap();

    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    let occupyable = area.scope::<Occupyable>()?.unwrap();
    let routes: Vec<Route> = occupyable.routes.clone().unwrap();
    assert_eq!(
        routes,
        vec![Route::Simple(SimpleRoute::new(
            "north",
            destination.entity_ref()
        ))]
    );

    build.close()?;

    Ok(())
}

#[test]
fn it_replaces_simple_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let previous = build.make(QuickThing::Place("Old Place"))?;
    let destination = build.make(QuickThing::Place("New Place"))?;
    let old_gid = previous.borrow().gid().unwrap();
    let new_gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", old_gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", new_gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;

    let (_, _person, area) = surroundings.unpack();

    let occupyable = area.scope::<Occupyable>()?.unwrap();
    let routes: Vec<Route> = occupyable.routes.clone().unwrap();
    assert_eq!(
        routes,
        vec![Route::Simple(SimpleRoute::new(
            "north",
            destination.entity_ref()
        ))]
    );

    build.close()?;

    Ok(())
}

#[test]
fn it_removes_simple_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;
    let (_, _person, area) = surroundings.unpack();

    let occupyable = area.scope::<Occupyable>()?.unwrap();
    let routes: Vec<Route> = occupyable.routes.clone().unwrap();
    assert_eq!(
        routes,
        vec![Route::Simple(SimpleRoute::new(
            "north",
            destination.entity_ref()
        ))]
    );

    let action = try_parsing(RouteActionParser {}, &format!("@route rm north"))?.unwrap();
    action.perform(session.clone(), &surroundings)?;
    let (_, _person, area) = surroundings.unpack();

    let occupyable = area.scope::<Occupyable>()?.unwrap();
    let routes: Vec<Route> = occupyable.routes.clone().unwrap();
    assert!(routes.is_empty());

    build.close()?;

    Ok(())
}

#[test]
fn it_navigates_through_simple_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;

    let action = try_parsing(GoActionParser {}, "go north")?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_picks_best_matching_route_by_name() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;
    let action = try_parsing(RouteActionParser {}, &format!("@route #{} south", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;
    let action = try_parsing(RouteActionParser {}, &format!("@route #{} northwest", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;

    let action = try_parsing(GoActionParser {}, "go east")?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    build.close()?;

    Ok(())
}

#[test]
fn it_deactivates_routes() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let gid = destination.borrow().gid().unwrap();
    let (session, surroundings) = build.build()?;

    let action = try_parsing(RouteActionParser {}, &format!("@route #{} north", gid))?.unwrap();
    action.perform(session.clone(), &surroundings)?;

    let action = DeactivateRouteAction {
        name: "north".to_owned(),
        reason: "A rock slide is blocking your way.".to_owned(),
    };
    action.perform(session.clone(), &surroundings)?;

    let action = try_parsing(GoActionParser {}, "go north")?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_goes_into_nested_areas() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Actual(destination.clone())])
        .build()?;

    let action = try_parsing(GoActionParser {}, "go place")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, area) = surroundings.unpack();

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &destination)?);

    assert_ne!(tools::area_of(&living)?.key(), area.key());
    assert_eq!(tools::area_of(&living)?.key(), destination.key());

    build.close()?;

    Ok(())
}
