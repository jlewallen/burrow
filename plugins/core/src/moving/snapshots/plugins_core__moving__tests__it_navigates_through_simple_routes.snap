---
source: plugins/core/src/moving/tests.rs
expression: reply.to_debug_json()?
---
{
  "reply": {
    "taggedJson": {
      "areaObservation": {
        "area": {
          "desc": null,
          "gid": 2,
          "key": "E-1",
          "name": "Place",
          "qualified": "Place"
        },
        "carrying": [],
        "items": [],
        "living": [
          {
            "desc": null,
            "gid": 3,
            "key": "E-2",
            "name": "Living",
            "qualified": "Living"
          }
        ],
        "person": {
          "desc": null,
          "gid": 3,
          "key": "E-2",
          "name": "Living",
          "qualified": "Living"
        },
        "routes": []
      }
    }
  }
}
